runtime/v2/manager.go

```diff
@@ -276,6 +287,79 @@ func (m *ShimManager) startShim(ctx context.Context, bundle *Bundle, id string,
 	return shim, nil
 }
 
+func (m *ShimManager) resolveRuntimePath(runtime string) (string, error) {
+	if runtime == "" {
+		return "", fmt.Errorf("no runtime name")
+	}
+
+	// Custom path to runtime binary
+	if filepath.IsAbs(runtime) {
+		// Make sure it exists before returning ok
+		if _, err := os.Stat(runtime); err != nil {
+			return "", fmt.Errorf("invalid custom binary path: %w", err)
+		}
+
+		return runtime, nil
+	}
+
+	// Preserve existing logic and resolve runtime path from runtime name.
+
+	name := shim_binary.BinaryName(runtime)
+	if name == "" {
+		return "", fmt.Errorf("invalid runtime name %s, correct runtime name should be either format like `io.containerd.runc.v1` or a full path to the binary", runtime)
+	}
+
+	if path, ok := m.runtimePaths.Load(name); ok {
```

Comment:
If understand it right, it should not be possible unless you move binaries while containerd is running?

Reply:
The runtime_path is part of task api. I think it is possible to use different path with same name from client side

Reply:
```
➜  containerd git:(review-6206.patch) sudo -E PATH=$PATH:/tmp bin/containerd > /tmp/6206.patch.log 2>&1 &
[1] 108696

➜  containerd git:(review-6206.patch) touch /tmp/containerd-shim-runc-v2

➜  containerd git:(review-6206.patch) sudo ctr c ls                                                                                  
CONTAINER    IMAGE    RUNTIME    

➜  containerd git:(review-6206.patch) sudo ctr run --rm --runtime io.containerd.runc.v2  docker.io/library/ubuntu:20.04 review-case ls
bin
boot
dev
etc
home
lib
lib32
lib64
libx32
media
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
➜  containerd git:(review-6206.patch) command -v containerd-shim-runc-v2
/usr/bin/containerd-shim-runc-v2

➜  containerd git:(review-6206.patch) sudo rm  $(command -v containerd-shim-runc-v2) 

➜  containerd git:(review-6206.patch) sudo ctr run --rm --runtime io.containerd.runc.v2  docker.io/library/ubuntu:20.04 review-case ls
ctr: failed to start shim: start failed: : fork/exec /usr/bin/containerd-shim-runc-v2: no such file or directory: unknown
```

Before this patch, the containerd can find other path for that runtime ( in the `/tmp`). I think the resolving runtime name in `PATH` env is better UX design. If user wants to use specific path, they should use absolute path feature provided by this patch.

But the cache locks the mapping between name and runtime path, which breaks the existing behavior. That is why I suggest to remove the cache.

Reply:
I can reproduce the problem you described in both `main` and in this PR.

Since containerd already has cached path for key `containerd-shim-runc-v2` and you remove the file, it'll fail with `no such file or directory`. We currently don't verify that cached file still exists.

This PR doesn't introduce new behavior, just moves the code to `manager.go`.

Have a look on the current main branch code:
https://github.com/containerd/containerd/blob/main/runtime/v2/shim/util.go#L73
https://github.com/containerd/containerd/blob/main/runtime/v2/shim/util.go#L109

<details>

$ git log --oneline -2
77e76d7a5 (HEAD, upstream/main, upstream/HEAD) Merge pull request #6246 from wzshiming/fix/close-file-with-modify-perm
eee8c969d (origin/main) Merge pull request #6250 from cpuguy83/fix_release_tar_name

$ make bin/containerd BUILDTAGS="no_btrfs"
+ bin/containerd


$ sudo -E PATH=$PATH:/tmp bin/containerd
…

$ stat /usr/local/bin/containerd-shim-runc-v2
  File: /usr/local/bin/containerd-shim-runc-v2
  Size: 9986048   	Blocks: 19504      IO Block: 4096   regular file


$ stat /tmp/containerd-shim-runc-v2 
  File: /tmp/containerd-shim-runc-v2
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file

$ sudo ctr run --rm --runtime io.containerd.runc.v2 docker.io/library/hello-world:latest test111

Hello from Docker!
This message shows that your installation appears to be working correctly.

To generate this message, Docker took the following steps:
 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
    (amd64)
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run an Ubuntu container with:
 $ docker run -it ubuntu bash

Share images, automate workflows, and more with a free Docker ID:
 https://hub.docker.com/

For more examples and ideas, visit:
 https://docs.docker.com/get-started/


$ sudo mv /usr/local/bin/containerd-shim-runc-v2 /usr/local/bin/OLD__containerd-shim-runc-v2

$ sudo ctr run --rm --runtime io.containerd.runc.v2 docker.io/library/hello-world:latest test111
ctr: failed to start shim: start failed: : fork/exec /usr/local/bin/containerd-shim-runc-v2: no such file or directory: unknown

</details>

Reply:
My mistake. I didn’t realize that the shim manager looks the binary in the current working dir at first and then in PATH. Since the working dir of bin/containerd is bin/ instead of /usr/bin, rm doesn’t impact the cache thing.

Reply:
This patch looks good to me. But I think we should have follow-up pr to remove that cache.